abi-versions/v0.2.1/README.md

```diff
@@ -0,0 +1,1996 @@
+
+If the named queue already exists, then it's going to be opened
+instead of creating a new empty queue.
+
+Items can be enqueued/dequeued on the created/opened queue using
+[`proxy_enqueue_shared_queue`] and/or [`proxy_dequeue_shared_queue`]
+with `return_queue_id`.
+
+Returned `status` value is:
+- `OK` on success.
+- `INVALID_MEMORY_ACCESS` when `name_data`, `name_size`
+  and/or `return_queue_id` point to invalid memory address.
+
+
+#### `proxy_resolve_shared_queue`
+
+* params:
+  - `i32 (const char *) vm_id_data`
+  - `i32 (size_t) vm_id_size`
+  - `i32 (const char *) name_data`
+  - `i32 (size_t) name_size`
+  - `i32 (uint32_t *) return_queue_id`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Resolves existing shared queue using the provided VM ID (`vm_id_data`,
+`vm_id_size`) and name (`name_data`, `name_size`).
+
+Items can be enqueued/dequeued on the opened queue using
+[`proxy_enqueue_shared_queue`] and/or [`proxy_dequeue_shared_queue`]
+with `return_queue_id`.
+
+Returned `status` value is:
+- `OK` on success.
+- `NOT_FOUND` when the requested queue was not found.
+- `INVALID_MEMORY_ACCESS` when `vm_id_data`, `vm_id_size`, `name_data`,
+  `name_size` and/or `return_queue_id` point to invalid memory address.
+
+
+#### `proxy_enqueue_shared_queue`
+
+* params:
+  - `i32 (uint32_t) queue_id`
+  - `i32 (const uint8_t *) value_data`
+  - `i32 (size_t) value_size`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Adds item (`value_data`, `value_size`) to the end of the queue
+`queue_id`.
+
+Returned `status` value is:
+- `OK` on success.
+- `NOT_FOUND` when the requested `queue_id` was not found.
+- `INVALID_MEMORY_ACCESS` when `value_data` and/or `value_size` point
+  to invalid memory address.
+
+
+#### `proxy_dequeue_shared_queue`
+
+* params:
+  - `i32 (uint32_t) queue_id`
+  - `i32 (uint8_t **) return_value_data`
+  - `i32 (size_t *) return_value_size`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Retrieves value (`return_value_data`, `return_value_size`) from
+the front of the queue `queue_id`.
+
+Returned `status` value is:
+- `OK` on success.
+- `NOT_FOUND` when the requested `queue_id` was not found.
+- `EMPTY` when there is nothing to dequeue from the requested queue.
+- `INVALID_MEMORY_ACCESS` when `return_value_data`
+  and/or `return_value_size` point to invalid memory address.
+
+
+### Callbacks exposed by the Wasm module
+
+#### `proxy_on_queue_ready`
+* params:
+  - `i32 (uint32_t) plugin_context_id`
+  - `i32 (uint32_t) queue_id`
+* returns:
+  - none
+
+Called when a new item is enqueued on the queue `queue_id`.
+
+
+## Metrics
+
+### Functions exposed by the host
+
+#### `proxy_define_metric`
+
+* params:
+  - `i32 (`[`proxy_metric_type_t`]`) metric_type`
+  - `i32 (const char *) name_data`
+  - `i32 (size_t) name_size`
+  - `i32 (uint32_t *) return_metric_id`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Defines a new metric of type `metric_type` with a name (`name_data`,
```

Comment:
A couple of questions about behaviors that are not specified here:

* What happens if I give it a metric name that was previously defined? Do I get the preexisting id in `return_metric_id`?
   * What happens if I pass a pre-existing name but with a different `metric_type`?
      * Are the names namespaced by type and I get a new metric id?
      * Do I get a `BAD_ARGUMENT` error?
      * Do I get the preexisting id and the type is ignored?
      * Do I get the preexisting id and the type is updated?

* This call creates a metric; how do I dispose of a metric? For other functions such as `proxy_set_property` setting the value to NULL can be assumed to mean "unset", but how to "undefine" and release memory from metrics?

Reply:
> * What happens if I give it a metric name that was previously defined? Do I get the preexisting id in `return_metric_id`?
>   
>   * What happens if I pass a pre-existing name but with a different `metric_type`?
>     
>     * Are the names namespaced by type and I get a new metric id?
>     * Do I get a `BAD_ARGUMENT` error?
>     * Do I get the preexisting id and the type is ignored?
>     * Do I get the preexisting id and the type is updated?

The current behavior (i.e. that matching Envoy) is that there can be different metric types that use the same metric name.

Also, Envoy returns a new `return_metric_id` for each call (even for the same metric name and type), but you should return the same metric ID for the same metric and type combination.

> * This call creates a metric; how do I dispose of a metric? For other functions such as `proxy_set_property` setting the value to NULL can be assumed to mean "unset", but how to "undefine" and release memory from metrics?

Removing of various resources (including metrics) is not well supported in v0.2.1 (which this PR tries to document). In the future, you'd use [proxy_delete_metric](https://github.com/PiotrSikora/spec/blob/review/abi-versions/vNEXT/README.md#proxy_delete_metric).